src/providers/ad/ad_machine_pw_renewal.c

```diff
@@ -390,20 +460,60 @@ errno_t ad_machine_account_password_renewal_init(struct be_ctx *be_ctx,
         offset = 0;
     }
 
-    ret = get_adcli_extra_args(dp_opt_get_cstring(ad_opts->basic, AD_DOMAIN),
+    renewal_data->renew_helper = RENEW_HELPER_REALM;
+    if (opt_list_size == 4) {
+        if (strcasecmp(opt_list[3], "adcli") == 0) {
+            renewal_data->renew_helper = RENEW_HELPER_ADCLI;
+        } else if (strcasecmp(opt_list[3], "realm") == 0) {
+            renewal_data->renew_helper = RENEW_HELPER_REALM;
+        } else {
+            DEBUG(SSSDBG_CRIT_FAILURE,
+                  "Unsupported keytab renewal helper program [%s].\n",
+                  opt_list[3]);
+            ret = EINVAL;
+            goto done;
+        }
+    }
+
+    if (renewal_data->renew_helper == RENEW_HELPER_ADCLI) {
+        ret = get_adcli_extra_args(dp_opt_get_cstring(ad_opts->basic, AD_DOMAIN),
                    dp_opt_get_cstring(ad_opts->basic, AD_HOSTNAME),
                    dp_opt_get_cstring(ad_opts->id_ctx->sdap_id_ctx->opts->basic,
                                       SDAP_KRB5_KEYTAB),
-                   dp_opt_get_bool(ad_opts->basic, AD_USE_LDAPS),
-                   lifetime,
+                   dp_opt_get_bool(ad_opts->basic, AD_USE_LDAPS), lifetime,
                    dp_opt_get_bool(ad_opts->basic,
                                    AD_UPDATE_SAMBA_MACHINE_ACCOUNT_PASSWORD),
                    period, initial_delay, renewal_data);
+    } else if (renewal_data->renew_helper == RENEW_HELPER_REALM) {
+        ret = get_realm_extra_args(dp_opt_get_cstring(ad_opts->basic, AD_DOMAIN),
+                   dp_opt_get_cstring(ad_opts->basic, AD_HOSTNAME),
+                   dp_opt_get_cstring(ad_opts->id_ctx->sdap_id_ctx->opts->basic,
+                                      SDAP_KRB5_KEYTAB),
+                   dp_opt_get_bool(ad_opts->basic, AD_USE_LDAPS), lifetime,
+                   dp_opt_get_bool(ad_opts->basic,
+                                   AD_UPDATE_SAMBA_MACHINE_ACCOUNT_PASSWORD),
+                   period, initial_delay, renewal_data);
+    } else {
+        DEBUG(SSSDBG_CRIT_FAILURE,
+              "Unsupported keytab renewal helper program.\n");
+        ret = EINVAL;
+        goto done;
+    }
     if (ret != EOK) {
         DEBUG(SSSDBG_OP_FAILURE, "get_adcli_extra_args failed.\n");
         goto done;
     }
 
+    ret = access(renewal_data->prog_path, X_OK);
+    if (ret != 0) {
+        ret = errno;
+        DEBUG(SSSDBG_CONF_SETTINGS,
+              "The helper program ["RENEWAL_PROG_PATH"] for renewal "
```

Comment:
It should be `renewal_data->prog_path`, not `RENEWAL_PROG_PATH`

Reply:
Fixed